Python/Product/VSCode/AnalysisVsc/LanguageServer.Lifetime.cs

```diff
@@ -0,0 +1,134 @@
+using StreamJsonRpc;
+
+namespace Microsoft.PythonTools.VsCode {
+    public partial class LanguageServer {
+        private TaskCompletionSource<InitializeResult> _tcs = new TaskCompletionSource<InitializeResult>();
+
+        [JsonRpcMethod("initialize")]
+        public Task<InitializeResult> Initialize(JToken token) {
+            var p = token.ToObject<InitializeParams>();
+            MonitorParentProcess(p);
+
+            _server.Initialize(p).ContinueWith(async t => {
+                if (t.IsCompletedSuccessfully) {
+                    await LoadDirectoryFiles(p);
+                    _tcs.SetResult(t.Result);
+                } else if (t.IsCanceled || _sessionTokenSource.IsCancellationRequested) {
+                    _tcs.TrySetCanceled();
+                } else if (t.Exception != null) {
+                    _tcs.TrySetException(t.Exception);
+                }
+            }).DoNotWait();
+
+            return _tcs.Task;
+        }
+
+        [JsonRpcMethod("initialized")]
+        public Task Initialized(JToken token)
+            => _server.Initialized(token.ToObject<InitializedParams>());
+
+        [JsonRpcMethod("shutdown")]
+        public Task Shutdown() => _server.Shutdown();
+
+        [JsonRpcMethod("exit")]
+        public async Task Exit() {
+            await _server.Exit();
+            _sessionTokenSource.Cancel();
+        }
+
+        private Task LoadDirectoryFiles(InitializeParams @params) {
+            string rootDir = null;
+            if (@params.rootUri != null) {
+                rootDir = @params.rootUri.ToAbsolutePath();
+            } else if (!string.IsNullOrEmpty(@params.rootPath)) {
+                rootDir = PathUtils.NormalizePath(@params.rootPath);
+            }
+
+            if (string.IsNullOrEmpty(rootDir)) {
+                return Task.CompletedTask;
+            }
+
+            var matcher = new Matcher();
+            matcher.AddIncludePatterns(@params.initializationOptions.includeFiles ?? new[] { "**/*" });
+            matcher.AddExcludePatterns(@params.initializationOptions.excludeFiles ?? Enumerable.Empty<string>());
+
+            var dib = new DirectoryInfoWrapper(new DirectoryInfo(rootDir));
+            var matchResult = matcher.Execute(dib);
+
+            _server.LogMessage(MessageType.Log, $"Loading files from {rootDir}");
+            return LoadFromDirectoryAsync(rootDir, matchResult);
+        }
+
+        private async Task LoadFromDirectoryAsync(string rootDir, PatternMatchingResult matchResult) {
+            foreach (var file in matchResult.Files) {
+                if (_sessionTokenSource.IsCancellationRequested) {
+                    break;
+                }
+
+                var path = Path.Combine(rootDir, file.Path);
+                if (!ModulePath.IsPythonSourceFile(path)) {
+                    if (ModulePath.IsPythonFile(path, true, true, true)) {
+                        // TODO: Deal with scrapable files (if we need to do anything?)
```

Comment:
I think the answer to this TODO is that when we resolve imports via search paths and the project root is passed as a search path then we'll import them just fine (which is not the default, but there's nothing stopping it - we could also just pass `rootDir` as a search path unconditionally, but I'm not sure that's always going to be right...).